src/adaptors/throttle.rs

```diff
@@ -251,75 +241,93 @@ async fn worker(limits: Limits, mut queue_rx: mpsc::Receiver<(Id, Sender<Never>)
 
         // as truncates which is ok since in case of truncation it would always be >=
         // limits.overall_s
-        let used = history
+        let used = when_requests_were_sent
             .iter()
             .take_while(|(_, time)| time > &sec_back)
             .count() as u32;
-        let mut allowed = limits.overall_s.saturating_sub(used);
+        let mut allowed = limits.messages_per_sec_overall.saturating_sub(used);
 
         if allowed == 0 {
-            hchats_s.clear();
+            requests_sent_to_chats.per_sec.clear();
             tokio::time::sleep(DELAY).await;
             continue;
         }
 
-        for (chat, _) in history.iter().take_while(|(_, time)| time > &sec_back) {
-            *hchats_s.entry(*chat).or_insert(0) += 1;
-        }
-
+        for (chat, _) in when_requests_were_sent
+            .iter()
+            .take_while(|(_, time)| time > &sec_back)
         {
-            let mut queue_rem = queue.removing();
-            while let Some(entry) = queue_rem.next() {
-                let chat = &entry.value().0;
-                let cond = {
-                    hchats_s.get(chat).copied().unwrap_or(0) < limits.chat_s
-                        && hchats.get(chat).copied().unwrap_or(0) < limits.chat_m
-                };
-
-                if cond {
-                    {
-                        *hchats_s.entry(*chat).or_insert(0) += 1;
-                        *hchats.entry(*chat).or_insert(0) += 1;
-                        history.push_back((*chat, Instant::now()));
-                    }
-
-                    // This will close the channel unlocking associated request
-                    drop(entry.remove());
+            *requests_sent_to_chats.per_sec.entry(*chat).or_insert(0) += 1;
+        }
 
-                    // We've "sent" 1 request, so now we can send 1 less
-                    allowed -= 1;
-                    if allowed == 0 {
-                        break;
-                    }
+        let mut queue_removing_iter = queue.removing();
```

Comment:
It's not really an iter though :thinking:

Reply:
Why this is not an iterator? We call `.next()` then.

Reply:
because it doesn't implement `Iterator`